src/Microsoft.OData.Core/Json/ODataJsonWriterUtils.cs

```diff
@@ -248,5 +256,89 @@ private static void WriteInnerError(IJsonWriter jsonWriter, ODataInnerError inne
             // }
             jsonWriter.EndObjectScope();
         }
+
+        internal static void ODataValueToString(StringBuilder sb, ODataValue value)
+        {
+            if (value == null || value is ODataNullValue)
+            {
+                sb.Append("null");
+            }
+
+            ODataCollectionValue collectionValue = value as ODataCollectionValue;
+            if (collectionValue != null)
+            {
+                ODataCollectionValueToString(sb, collectionValue);
+            }
+
+            ODataResourceValue resourceValue = value as ODataResourceValue;
+            if (resourceValue != null)
+            {
+                ODataResourceValueToString(sb, resourceValue);
+            }
+
+            ODataPrimitiveValue primitiveValue = value as ODataPrimitiveValue;
+            if (primitiveValue != null)
+            {
+                if (primitiveValue.FromODataValue() is string)
+                {
+                    sb.Append("\"" + JsonValueUtils.GetEscapedJsonString(value.FromODataValue()?.ToString()) + "\"");
+                }
+                else
+                {
+                    sb.Append(JsonValueUtils.GetEscapedJsonString(value.FromODataValue()?.ToString()));
+                }
+            }
+        }
+
+        private static void ODataCollectionValueToString(StringBuilder sb, ODataCollectionValue value)
+        {
+            bool isFirst = true;
+            sb.Append("[");
+            foreach (object item in value.Items)
+            {
+                if (isFirst)
+                {
+                    isFirst = false;
+                }
+                else
+                {
+                    sb.Append(",");
+                }
+
+                ODataValue odataValue = item as ODataValue;
+                if (odataValue != null)
+                {
+                    ODataValueToString(sb, odataValue);
+                }
+                else

```

Comment:
Things start getting complicated soon. Like for collection of guids, how do you read those? I think it is fine to only allow for ODataValues. Since it is not going to be very frequently used, I think this is fine. 

---
In reply to: [323952068](https://github.com/OData/odata.net/pull/1523#discussion_r323952068) [](ancestors = 323952068)